### PR TITLE
927 go 122 upgrade check list

### LINKED ISSUE #927: Go 1.22 upgrade check list
See https://github.com/nspcc-dev/neo-go/issues/3310

### PR DESCRIPTION
**Pull request by smallhive**
Closes #927

### CODE REVIEW

File: api/auth/center_test.go [hunk 1]
```diff
@@ -318,7 +318,7 @@ func TestAwsEncodedChunkReader(t *testing.T) {
 	t.Run("err chunk header too long", func(t *testing.T) {
 		streamSigner := v4.NewChunkSigner("us-east-1", "s3", seedSignature, ts, awsCreds)
 		chunkThreeBody := make([]byte, 4097)
-		for i := 0; i < len(chunkThreeBody); i++ {
+		for i := range len(chunkThreeBody) {
```
Comment (roman-khimov) [inline api/auth/center_test.go:322]: `range chunkThreeBody`

File: api/handler/encryption_test.go [hunk 2]
```diff
@@ -120,7 +120,7 @@ func equalDataSlices(t *testing.T, expected, actual []byte) {
 		return
 	}
 
-	for i := 0; i < len(expected); i++ {
+	for i := range len(expected) {
```
Comment (roman-khimov) [inline api/handler/encryption_test.go:124]: `slices.Equal`?

File: api/handler/object_list_test.go [hunk 3]
```diff
@@ -154,12 +154,12 @@ func validateListV2(t *testing.T, tc *handlerContext, bktName, prefix, delimiter
...
 
 	require.Len(t, response.Contents, len(checkObjects))
-	for i := 0; i < len(checkObjects); i++ {
+	for i := range len(checkObjects) {
 		require.Equal(t, checkObjects[i], response.Contents[i].Key)
 	}
 
 	require.Len(t, response.CommonPrefixes, len(checkPrefixes))
-	for i := 0; i < len(checkPrefixes); i++ {
+	for i := range len(checkPrefixes) {
```
Comment (roman-khimov) [inline api/handler/object_list_test.go:162]: drop `len`

File: api/handler/object_list_test.go [hunk 4]
```diff
@@ -154,12 +154,12 @@ func validateListV2(t *testing.T, tc *handlerContext, bktName, prefix, delimiter
 	require.Equal(t, last, len(response.NextContinuationToken) == 0)
 
 	require.Len(t, response.Contents, len(checkObjects))
-	for i := 0; i < len(checkObjects); i++ {
+	for i := range len(checkObjects) {
```
Comment (roman-khimov) [inline api/handler/object_list_test.go:157]: drop `len`

File: api/handler/s3encoder.go [hunk 5]
```diff
@@ -32,7 +32,7 @@ func shouldEscape(c byte) bool {
 // while considering some S3 exceptions.
 func s3URLEncode(s string, mode encoding) string {
 	spaceCount, hexCount := 0, 0
-	for i := 0; i < len(s); i++ {
+	for i := range len(s) {
```
Comment (roman-khimov) [inline api/handler/s3encoder.go:35]: `range s`

File: api/handler/s3encoder.go [hunk 6]
```diff
@@ -59,7 +59,7 @@ func s3URLEncode(s string, mode encoding) string {
 
 	if hexCount == 0 {
 		copy(t, s)
-		for i := 0; i < len(s); i++ {
+		for i := range len(s) {
```
Comment (roman-khimov) [inline api/handler/s3encoder.go:62]: `range s`

File: api/handler/s3encoder.go [hunk 7]
```diff
@@ -68,7 +68,7 @@ func s3URLEncode(s string, mode encoding) string {
 	}
 
 	j := 0
-	for i := 0; i < len(s); i++ {
+	for i := range len(s) {
```
Comment (roman-khimov) [inline api/handler/s3encoder.go:71]: `range s`

File: api/auth/center_test.go [hunk 8]
```diff
@@ -115,7 +116,7 @@ func TestAwsEncodedChunkReader(t *testing.T) {
 		chunkOnePayload[i] = 'a'
 	}
 
-	chunkOneBody := append([]byte("10000;chunk-signature=ad80c730a21e5b8d04586a2213dd63b9a0e99e0e2307b0ade35a65485a288648\n"), chunkOnePayload...)
+	chunkOneBody := slices.Concat([]byte("10000;chunk-signature=ad80c730a21e5b8d04586a2213dd63b9a0e99e0e2307b0ade35a65485a288648\n"), chunkOnePayload)
```
Comment (roman-khimov) [inline api/auth/center_test.go:119]: This better be left as is, we've got a new slice inside anyway.

File: api/auth/center_test.go [hunk 9]
```diff
@@ -129,7 +130,7 @@ func TestAwsEncodedChunkReader(t *testing.T) {
 		chunkTwoPayload[i] = 'a'
 	}
 
-	chunkTwoBody := append([]byte("400;chunk-signature=0055627c9e194cb4542bae2aa5492e3c1575bbb81b612b7d234b86a503ef5497\n"), chunkTwoPayload...)
+	chunkTwoBody := slices.Concat([]byte("400;chunk-signature=0055627c9e194cb4542bae2aa5492e3c1575bbb81b612b7d234b86a503ef5497\n"), chunkTwoPayload)
```
Comment (roman-khimov) [inline api/auth/center_test.go:133]: Also.

File: api/auth/center_test.go [hunk 10]
```diff
@@ -404,10 +405,10 @@ func TestAwsEncodedWithRequest(t *testing.T) {
 			body = []byte{'\r', '\n'}
 		}
 
-		body = append(body, []byte(strconv.FormatInt(int64(len(chunkPayload)), 16)+";chunk-signature=")...)
-		body = append(body, []byte(hex.EncodeToString(chunkSignature))...)
+		body = slices.Concat(body, []byte(strconv.FormatInt(int64(len(chunkPayload)), 16)+";chunk-signature="))
+		body = slices.Concat(body, []byte(hex.EncodeToString(chunkSignature)))
 		body = append(body, '\n')
-		body = append(body, chunkPayload...)
+		body = slices.Concat(body, chunkPayload)
```
Comment (roman-khimov) [inline api/auth/center_test.go:411]: We're assembling some `body` here, I doubt `slices.Concat` is appropriate.

File: api/auth/center_test.go [hunk 11]
```diff
@@ -417,7 +418,7 @@ func TestAwsEncodedWithRequest(t *testing.T) {
 	signChunk, err := streamSigner.GetSignature(nil)
 	require.NoError(t, err)
 
-	chunk3Body := append([]byte("\r\n0;chunk-signature="), []byte(hex.EncodeToString(signChunk))...)
+	chunk3Body := slices.Concat([]byte("\r\n0;chunk-signature="), []byte(hex.EncodeToString(signChunk)))
```
Comment (roman-khimov) [inline api/auth/center_test.go:421]: Not needed.

File: api/auth/signer/v4/v4.go [hunk 12]
```diff
@@ -630,7 +630,7 @@ func (ctx *signingCtx) buildCanonicalHeaders(r rule, header http.Header) {
 		lowerCaseKey := strings.ToLower(k)
 		if _, ok := ctx.SignedHeaderVals[lowerCaseKey]; ok {
 			// include additional values
-			ctx.SignedHeaderVals[lowerCaseKey] = append(ctx.SignedHeaderVals[lowerCaseKey], v...)
+			ctx.SignedHeaderVals[lowerCaseKey] = slices.Concat(ctx.SignedHeaderVals[lowerCaseKey], v)
```
Comment (roman-khimov) [inline api/auth/signer/v4/v4.go:633]: Do we need a new slice to be allocated here?

File: api/handler/acl.go [hunk 13]
```diff
@@ -841,7 +841,7 @@ func getAstOps(resource *astResource, childOp *astOperation) []*astOperation {
 func removeAstOp(resource *astResource, group bool, op eacl.Operation, action eacl.Action) {
 	for i, astOp := range resource.Operations {
 		if astOp.IsGroupGrantee() == group && astOp.Op == op && astOp.Action == action {
-			resource.Operations = append(resource.Operations[:i], resource.Operations[i+1:]...)
+			resource.Operations = slices.Concat(resource.Operations[:i], resource.Operations[i+1:])
```
Comment (roman-khimov) [inline api/handler/acl.go:844]: This looks more like a `slices.Delete`.

File: api/handler/acl.go [hunk 14]
```diff
@@ -850,7 +850,7 @@ func removeAstOp(resource *astResource, group bool, op eacl.Operation, action ea
 func addUsers(resource *astResource, astO *astOperation, users []user.ID) {
 	for _, astOp := range resource.Operations {
 		if astOp.IsGroupGrantee() == astO.IsGroupGrantee() && astOp.Op == astO.Op && astOp.Action == astO.Action {
-			astOp.Users = append(astO.Users, users...)
+			astOp.Users = slices.Concat(astO.Users, users)
```
Comment (roman-khimov) [inline api/handler/acl.go:853]: Also not sure if an additional allocation is appropriate.

File: api/handler/acl.go [hunk 15]
```diff
@@ -866,7 +866,7 @@ func removeUsers(resource *astResource, astOperation *astOperation, users []user
 				}
 			}
 			if len(filteredUsers) == 0 { // remove ast resource
-				resource.Operations = append(resource.Operations[:ind], resource.Operations[ind+1:]...)
+				resource.Operations = slices.Concat(resource.Operations[:ind], resource.Operations[ind+1:])
```
Comment (roman-khimov) [inline api/handler/acl.go:869]: Deletion.

File: api/handler/acl.go [hunk 16]
```diff
@@ -1002,7 +1002,7 @@ func addToList(operations []*astOperation, rec eacl.Record, target eacl.Target)
 			Action: rec.Action(),
 		}
 		if !groupTarget {
-			astOperation.Users = append(astOperation.Users, target.Accounts()...)
+			astOperation.Users = slices.Concat(astOperation.Users, target.Accounts())
```
Comment (roman-khimov) [inline api/handler/acl.go:1005]: Unwanted allocation.

File: api/handler/acl.go [hunk 17]
```diff
@@ -1268,7 +1268,7 @@ func aclToPolicy(acl *AccessControlPolicy, resInfo *resourceInfo) (*bucketPolicy
 	// parseACLHeaders(). If there is no other grants, then user sets private
 	// canned ACL, which is processed in this branch.
 	if len(acl.AccessControlList) < 2 {
-		results = append([]statement{getDenyStatement(resInfo, allUsersWildcard, awsPermFullControl)}, results...)
+		results = slices.Concat([]statement{getDenyStatement(resInfo, allUsersWildcard, awsPermFullControl)}, results)
```
Comment (roman-khimov) [inline api/handler/acl.go:1271]: Not needed.

File: api/handler/encryption_test.go [hunk 18]
```diff
@@ -161,7 +162,7 @@ func multipartUploadEncrypted(hc *handlerContext, bktName, objName string, heade
 
 		etag, data := uploadPartEncrypted(hc, bktName, objName, multipartInfo.UploadID, currentPart, adjustedSize)
 		etags = append(etags, etag)
-		objData = append(objData, data...)
+		objData = slices.Concat(objData, data)
```
Comment (roman-khimov) [inline api/handler/encryption_test.go:165]: Inappropriate.

File: api/handler/util.go [hunk 19]
```diff
@@ -25,7 +26,7 @@ func (h *handler) logAndSendError(w http.ResponseWriter, logText string, reqInfo
 		zap.String("object", reqInfo.ObjectName),
 		zap.String("description", logText),
 		zap.Error(err)}
-	fields = append(fields, additional...)
+	fields = slices.Concat(fields, additional)
```
Comment (roman-khimov) [inline api/handler/util.go:29]: Inappropriate.

File: api/layer/neofs_mock.go [hunk 20]
```diff
@@ -300,7 +301,7 @@ func (t *TestNeoFS) CreateObject(_ context.Context, prm PrmObjectCreate) (oid.ID
 			for _, e := range prm.Multipart.Link.Objects() {
 				addr = newAddress(prm.Container, e.ObjectID())
 				if partialObject, ok := t.objects[addr.EncodeToString()]; ok {
-					payload = append(payload, partialObject.Payload()...)
+					payload = slices.Concat(payload, partialObject.Payload())
```
Comment (roman-khimov) [inline api/layer/neofs_mock.go:304]: Inappropriate.

File: api/layer/tree_mock.go [hunk 21]
```diff
@@ -251,7 +251,7 @@ func (t *TreeServiceMock) RemoveVersion(_ context.Context, bktInfo *data.BucketI
 	for key, versions := range cnrVersionsMap {
 		for i, node := range versions {
 			if node.ID == nodeID {
-				cnrVersionsMap[key] = append(versions[:i], versions[i+1:]...)
+				cnrVersionsMap[key] = slices.Concat(versions[:i], versions[i+1:])
```
Comment (roman-khimov) [inline api/layer/tree_mock.go:254]: Deletion.

File: api/layer/tree_mock.go [hunk 22]
```diff
@@ -468,7 +468,7 @@ LOOP:
 		for i, multipart := range multiparts {
 			if multipart.ID == multipartNodeID {
 				uploadID = multipart.UploadID
-				cnrMultipartsMap[key] = append(multiparts[:i], multiparts[i+1:]...)
+				cnrMultipartsMap[key] = slices.Concat(multiparts[:i], multiparts[i+1:])
```
Comment (roman-khimov) [inline api/layer/tree_mock.go:471]: Deletion.

File: api/reqinfo.go [hunk 23]
```diff
@@ -182,7 +183,7 @@ func (r *ReqInfo) GetTags() []KeyVal {
 	}
 	r.RLock()
 	defer r.RUnlock()
-	return append([]KeyVal(nil), r.tags...)
+	return slices.Concat([]KeyVal(nil), r.tags)
```
Comment (roman-khimov) [inline api/reqinfo.go:186]: This looks like a `Clone()`.

File: internal/neofs/tree.go [hunk 24]
```diff
@@ -686,7 +686,7 @@ func (c *TreeClient) getSubTreeByPrefix(ctx context.Context, bktInfo *data.Bucke
 		} else if !latestOnly || isIntermediate(node) {
 			nodes = append(nodes, node)
 		} else if isIntermediate(nodes[0]) {
-			nodes = append([]*tree.GetSubTreeResponse_Body{node}, nodes...)
+			nodes = slices.Concat([]*tree.GetSubTreeResponse_Body{node}, nodes)
```
Comment (roman-khimov) [inline internal/neofs/tree.go:689]: `append` is better here.

File: internal/neofs/tree.go [hunk 25]
```diff
@@ -696,7 +696,7 @@ func (c *TreeClient) getSubTreeByPrefix(ctx context.Context, bktInfo *data.Bucke
 
 	result := make([]*tree.GetSubTreeResponse_Body, 0, len(subTree))
 	for _, nodes := range nodesMap {
-		result = append(result, nodes...)
+		result = slices.Concat(result, nodes)
```
Comment (roman-khimov) [inline internal/neofs/tree.go:699]: `append` wins, it's even preallocated for that.

File: internal/neofs/tree.go [hunk 26]
```diff
@@ -773,7 +773,7 @@ func (c *TreeClient) getSubTreeVersions(ctx context.Context, bktInfo *data.Bucke
 		if latestOnly && version[0].IsDeleteMarker() {
 			continue
 		}
-		result = append(result, version...)
+		result = slices.Concat(result, version)
```
Comment (roman-khimov) [inline internal/neofs/tree.go:776]: Also

File: internal/neofs/tree.go [hunk 27]
```diff
@@ -825,7 +825,7 @@ func (c *TreeClient) getVersionsByPrefix(ctx context.Context, bktInfo *data.Buck
 		if err != nil {
 			return nil, err
 		}
-		result = append(result, versions...)
+		result = slices.Concat(result, versions)
```
Comment (roman-khimov) [inline internal/neofs/tree.go:828]: Also

File: internal/neofs/tree.go [hunk 28]
```diff
@@ -878,7 +878,7 @@ func (c *TreeClient) GetMultipartUploadsByPrefix(ctx context.Context, bktInfo *d
 		if err != nil {
 			return nil, err
 		}
-		result = append(result, multipartUploads...)
+		result = slices.Concat(result, multipartUploads)
```
Comment (roman-khimov) [inline internal/neofs/tree.go:881]: Also.

File: go.mod [hunk 29]
```diff
@@ -18,10 +18,10 @@ require (
 	github.com/spf13/pflag v1.0.5
 	github.com/spf13/viper v1.19.0
 	github.com/stretchr/testify v1.9.0
-	github.com/urfave/cli/v2 v2.27.2
+	github.com/urfave/cli/v2 v2.27.4
```
Comment (roman-khimov) [inline go.mod:21]: Wow. Surprised to see it here.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: api/auth/center_test.go [hunk 30]
```diff
@@ -432,7 +433,7 @@ func TestAwsEncodedWithRequest(t *testing.T) {
 func chunkSlice(payload []byte, chunkSize int) [][]byte {
 	var result [][]byte
 
-	for i := 0; i < len(payload); i += chunkSize {
+	for i := range len(payload) {
```
Comment (roman-khimov) [inline api/auth/center_test.go:436]: This doesn't seem to be correct, we were iterating with `+= chunkSize` steps.

**Review by roman-khimov — COMMENTED** [review]
Now `slices.Concat` commit is not just about `Concat`.

File: api/auth/center_test.go (hunk 30 above)
Comment (smallhive) [reply api/auth/center_test.go]: Damn, updated

**Comment by smallhive** [comment]
> Now `slices.Concat` commit is not just about `Concat`.

Renamed commit text

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**